init blog intro structured logs (#23632)
kubernetes/website: content/en/blog/_posts/2020-09-04-introducing-structured-logs.md

---
layout: blog 
title: 'Introducing Structured Logs'
date: 2020-09-04 

> slug: kubernetes-1-19-Introducing-Structured-Logs

Reviewer: slug: kubernetes-1-19-introducing-structured-logs

---

**Authors:** Marek Siarkowicz (Google), Nathan Beach (Google)

Logs are an essential aspect of observability and a critical tool for debugging. But Kubernetes logs have traditionally been unstructured strings, making any automated parsing difficult and any downstream processing, analysis, or querying challenging to do reliably.

> In Kubernetes 1.19, we are adding support for structured logs, which natively support (key, value) pairs and object references. We have also updated many logging calls such that over 99% of logging volume in a typical deployment are now migrated to the structured format.

Reviewer: In Kubernetes 1.19, we are adding support for [structured logs](/docs/concepts/cluster-administration/system-logs/), which natively support (key, value) pairs and object references.

> To maintain backwards compatibility, structured logs will still be outputted as a string where the string contains representations of those "key"="value" pairs. Starting in alpha in 1.19, logs can also be outputted in JSON format using the `--logging-format=json` flag.

Reviewer 2: ```suggestion
To maintain backwards compatibility, structured logs will still be output as a string where the string contains representations of the log's "key"="value" pairs. Starting in alpha in 1.19, logs can also be output in JSON using the `--logging-format=json` flag.
```

## Using Structured Logs

> We've added two new methods to the klog library: InfoS and ErrorS. For example, this invocation of InfoS:

Reviewer: We've added two new methods to the klog library: `InfoS` and `ErrorS`. For example, this invocation of `InfoS`:

```golang
klog.InfoS(“Pod status updated”, “pod”, klog.KObj(pod), “status”, status)
```

will result in this log:

```
I1025 00:15:15.525108       1 controller_utils.go:116] "Pod status updated" pod="kube-system/kubedns" status=”ready”
```

> Or, if the --logging-format=json flag is set, it will result in this output:

Reviewer: Or, if the `--logging-format=json` flag is set, it will result in this output:

```json
{
  "ts": 1580306777.04728,
  "msg": "Pod status updated",
  "Pod”: {
    “Name”: “kubedns”,
    “Namespace”: “kubedns”
  },
  “status”: “ready”
}
```

> This means downstream logging tools can easily ingest structured logging data and no longer need to maintain brittle regular expressions in an attempt to parse unstructured strings. This also makes processing logs easier, querying logs more robust, and analyzing logs much faster.

Reviewer 2: ```suggestion
This means downstream logging tools can easily ingest structured logging data and instead of using regular expressions to parse unstructured strings. This also makes processing logs easier, querying logs more robust, and analyzing logs much faster.
```

> With structured logs, all references to Kubernetes objects are structured the same way, so it's straightforward to filter to only log entries referencing the particular pod. You can easily find logs indicating how the scheduler was scheduling the pod, how the pod was created, the health probes of the pod, and all other changes in the lifecycle of the pod.

Reviewer 2: ```suggestion
With structured logs, all references to Kubernetes objects are structured the same way, so you can filter the output and only log entries referencing the particular pod. You can also find logs indicating how the scheduler was scheduling the pod, how the pod was created, the health probes of the pod, and all other changes in the lifecycle of the pod.
```

(Avoid words like 'easy' and 'straightforward', what's easy for you might not be easy for someone else)

> Suppose you are debugging an issue with a pod. With structured logs, you can easily filter to only those log entries referencing the pod of interest -- rather than needing to scan through potentially thousands of log lines to find the relevant ones.

Reviewer 2: ```suggestion
Suppose you are debugging an issue with a pod. With structured logs, you can filter to only those log entries referencing the pod of interest, rather than needing to scan through potentially thousands of log lines to find the relevant ones.
```

> Not only are structured logs easier for manual debugging of issues, they also enable richer features like automated pattern recognition within logs or tighter correlation of log and trace data.

Reviewer 2: ```suggestion
Not only are structured logs more useful when manual debugging of issues, they also enable richer features like automated pattern recognition within logs or tighter correlation of log and trace data.
```
